samples/bluetooth/alexa_gadget/src/gadgets/CMakeLists.txt

```diff
@@ -0,0 +1,197 @@
+    message(STATUS "and extract to either this sample directory or somewhere cmake can find it via system PATH.")
+    message(FATAL_ERROR "nanopb required by this project.")
+endif()
+
+message(STATUS "Using protoc: ${PROTOC_BIN}")
+message(STATUS "Using protoc-gen-nanopb: ${NANOPB_BIN}")
+
+if(ALEXA_EMBEDDED_PATH)
+  # Do nothing, just use the provided path
+elseif(WEST)
+  ## Use Python and west to find the alexa-embedded path
+  execute_process(
+    COMMAND ${PYTHON_EXECUTABLE} -c
+    "from west.manifest import Manifest; manifest = Manifest.from_file(); print(manifest.get_projects(['Alexa-Gadgets-Embedded-Sample-Code'])[0].posixpath)"
+    OUTPUT_VARIABLE
+    west_project_output
+    RESULT_VARIABLE
+    result
+  )
+  string(REGEX REPLACE "[\r\n]+" "" ALEXA_EMBEDDED_PATH "${west_project_output}")
+  if(${result})
+    message(FATAL_ERROR "Failed to find alexa-embedded, cannot build sample")
+  endif()
+else()
+  message(FATAL_ERROR "west not installed, please provide ALEXA_EMBEDDED_PATH to CMake to build this sample")
+endif()
+
+if(NANOPB_PATH)
+  # Do nothing, just use the provided path
+elseif(WEST)
+  ## Use Python and west to find the nanopb path
+  execute_process(
+    COMMAND ${PYTHON_EXECUTABLE} -c
+     "from west.manifest import Manifest; manifest = Manifest.from_file(); print(manifest.get_projects(['nanopb'])[0].posixpath)"
+    OUTPUT_VARIABLE
+    west_project_output
+    RESULT_VARIABLE
+    result
+  )
+  string(REGEX REPLACE "[\r\n]+" "" NANOPB_PATH "${west_project_output}")
+  if(${result})
+    message(FATAL_ERROR "Failed to find nanopb source files, cannot build sample")
+  endif()
+else()
+  message(FATAL_ERROR "west not installed, please provide NANOPB_PATH to CMake to build this sample")
+endif()
+
+function(compile_protobufs_in_dir dir recursive)
+    # Find all .proto files and include directories
+    if (recursive)
+        file(GLOB_RECURSE PROTO_FILES "${dir}/*.proto")
+    else()
+        file(GLOB PROTO_FILES "${dir}/*.proto")
+    endif()
+
+    foreach(file ${PROTO_FILES})
+        get_filename_component(FILE_PATH ${file} DIRECTORY)
+        list(APPEND PROTOC_INC_DIRS "-I${FILE_PATH}")
+    endforeach()
+
+    # Compile .proto files in their respective directories
+    foreach(file ${PROTO_FILES})
+        # Add all directories containing .proto files.
+        # Duplicates and current directory is then removed from "-I" arguments
+        set(PROTOC_ARGS "${PROTOC_INC_DIRS}")
+        get_filename_component(FILE_PATH ${file} DIRECTORY)
+        get_filename_component(FILE_NAME ${file} NAME)
+        list(REMOVE_ITEM PROTOC_ARGS "-I${FILE_PATH}")
+        list(REMOVE_DUPLICATES PROTOC_ARGS)
+        list(APPEND PROTOC_ARGS "-I.")
+        list(APPEND PROTOC_ARGS "--nanopb_out=${FILE_PATH}")
+        list(APPEND PROTOC_ARGS "${FILE_NAME}")
+        message(STATUS "Compiling proto definition: ${FILE_NAME}")
+        execute_process(
+            COMMAND ${PROTOC_BIN} ${PROTOC_ARGS}
```

Comment:
This creates a lot of untracked files in the repo:
```
HEAD detached at refs/heads/manifest-rev
Untracked files:
  (use "git add <file>..." to include in what will be committed)

        AlexaGadgetsProtobuf/Alerts/DeleteAlert/alertsDeleteAlertDirective.pb.c
        AlexaGadgetsProtobuf/Alerts/DeleteAlert/alertsDeleteAlertDirective.pb.h
        AlexaGadgetsProtobuf/Alerts/DeleteAlert/alertsDeleteAlertDirectivePayload.pb.c
        AlexaGadgetsProtobuf/Alerts/DeleteAlert/alertsDeleteAlertDirectivePayload.pb.h
        AlexaGadgetsProtobuf/Alerts/SetAlert/alertsSetAlertDirective.pb.c
        AlexaGadgetsProtobuf/Alerts/SetAlert/alertsSetAlertDirective.pb.h
        AlexaGadgetsProtobuf/Alerts/SetAlert/alertsSetAlertDirectivePayload.pb.c
        AlexaGadgetsProtobuf/Alerts/SetAlert/alertsSetAlertDirectivePayload.pb.h
        AlexaGadgetsProtobuf/Alexa.Discovery/Discover.Response/alexaDiscoveryDiscoverResponseEvent.pb.c
        AlexaGadgetsProtobuf/Alexa.Discovery/Discover.Response/alexaDiscoveryDiscoverResponseEvent.pb.h
        AlexaGadgetsProtobuf/Alexa.Discovery/Discover.Response/alexaDiscoveryDiscoverResponseEventPayload.pb.c
        AlexaGadgetsProtobuf/Alexa.Discovery/Discover.Response/alexaDiscoveryDiscoverResponseEventPayload.pb.h
        AlexaGadgetsProtobuf/Alexa.Discovery/Discover/alexaDiscoveryDiscoverDirective.pb.c
        AlexaGadgetsProtobuf/Alexa.Discovery/Discover/alexaDiscoveryDiscoverDirective.pb.h
        AlexaGadgetsProtobuf/Alexa.Discovery/Discover/alexaDiscoveryDiscoverDirectivePayload.pb.c
        AlexaGadgetsProtobuf/Alexa.Discovery/Discover/alexaDiscoveryDiscoverDirectivePayload.pb.h
        AlexaGadgetsProtobuf/Alexa.Gadget.MusicData/Tempo/alexaGadgetMusicDataTempoDirective.pb.c
        AlexaGadgetsProtobuf/Alexa.Gadget.MusicData/Tempo/alexaGadgetMusicDataTempoDirective.pb.h
        AlexaGadgetsProtobuf/Alexa.Gadget.MusicData/Tempo/alexaGadgetMusicDataTempoDirectivePayload.pb.c
        AlexaGadgetsProtobuf/Alexa.Gadget.MusicData/Tempo/alexaGadgetMusicDataTempoDirectivePayload.pb.h
        AlexaGadgetsProtobuf/Alexa.Gadget.SpeechData/Speechmarks/alexaGadgetSpeechDataSpeechmarksDirective.pb.c
        AlexaGadgetsProtobuf/Alexa.Gadget.SpeechData/Speechmarks/alexaGadgetSpeechDataSpeechmarksDirective.pb.h
        AlexaGadgetsProtobuf/Alexa.Gadget.SpeechData/Speechmarks/alexaGadgetSpeechDataSpeechmarksDirectivePayload.pb.c
        AlexaGadgetsProtobuf/Alexa.Gadget.SpeechData/Speechmarks/alexaGadgetSpeechDataSpeechmarksDirectivePayload.pb.h
        AlexaGadgetsProtobuf/Alexa.Gadget.StateListener/StateUpdate/alexaGadgetStateListenerStateUpdateDirective.pb.c
        AlexaGadgetsProtobuf/Alexa.Gadget.StateListener/StateUpdate/alexaGadgetStateListenerStateUpdateDirective.pb.h
        AlexaGadgetsProtobuf/Alexa.Gadget.StateListener/StateUpdate/alexaGadgetStateListenerStateUpdateDirectivePayload.pb.c
        AlexaGadgetsProtobuf/Alexa.Gadget.StateListener/StateUpdate/alexaGadgetStateListenerStateUpdateDirectivePayload.pb.h
        AlexaGadgetsProtobuf/Notifications/ClearIndicator/notificationsClearIndicatorDirective.pb.c
        AlexaGadgetsProtobuf/Notifications/ClearIndicator/notificationsClearIndicatorDirective.pb.h
        AlexaGadgetsProtobuf/Notifications/ClearIndicator/notificationsClearIndicatorDirectivePayload.pb.c
        AlexaGadgetsProtobuf/Notifications/ClearIndicator/notificationsClearIndicatorDirectivePayload.pb.h
        AlexaGadgetsProtobuf/Notifications/SetIndicator/notificationsSetIndicatorDirective.pb.c
        AlexaGadgetsProtobuf/Notifications/SetIndicator/notificationsSetIndicatorDirective.pb.h
        AlexaGadgetsProtobuf/Notifications/SetIndicator/notificationsSetIndicatorDirectivePayload.pb.c
        AlexaGadgetsProtobuf/Notifications/SetIndicator/notificationsSetIndicatorDirectivePayload.pb.h
        AlexaGadgetsProtobuf/common/directiveHeader.pb.c
        AlexaGadgetsProtobuf/common/directiveHeader.pb.h
        AlexaGadgetsProtobuf/common/eventHeader.pb.c
        AlexaGadgetsProtobuf/common/eventHeader.pb.h
        AlexaGadgetsProtobuf/parser/directiveParser.pb.c
        AlexaGadgetsProtobuf/parser/directiveParser.pb.h
        AlexaGadgetsProtobuf/parser/eventParser.pb.c
        AlexaGadgetsProtobuf/parser/eventParser.pb.h
        ConnectionHelpers/BLE/Proto/accessories.pb.c
        ConnectionHelpers/BLE/Proto/accessories.pb.h
        ConnectionHelpers/BLE/Proto/common.pb.c
        ConnectionHelpers/BLE/Proto/common.pb.h
        ConnectionHelpers/BLE/Proto/device.pb.c
        ConnectionHelpers/BLE/Proto/device.pb.h
        ConnectionHelpers/BLE/Proto/firmware.pb.c
        ConnectionHelpers/BLE/Proto/firmware.pb.h
```

Generated files must be placed in the build folder.

Reply:
also, we should not compile source file at the CMake configure stage, please use `add_custom_target` and mark those files as `BYPRODUCTS`, that will ensure that CMake and Ninja/make can handle those files correctly.

https://cmake.org/cmake/help/latest/command/add_custom_target.html

The file used for later build stages will automatically be marked with generated property.
https://cmake.org/cmake/help/latest/prop_sf/GENERATED.html

Doing so has a couple of benefits:
- Dependencies are set up correctly, so a change to a proto file results in a rebuild
- `clean` target will be able to remove generated files.
- CMake does not compile files, ensuring logical split between configure, generator, and build stages.